--- oscar/scripts/oscar_wizard	2007-07-28 00:39:02.000000000 -0400
+++ oscar/scripts/oscar_wizard.new	2007-08-16 13:07:21.000000000 -0400
@@ -170,7 +170,7 @@
 #
 # The title Group
 #
-my $image = "$ENV{OSCAR_HOME}/images/oscar.gif"; 
+my $image = "/usr/share/oscar/images/oscar.gif"; 
 
 my $GIF_t=$window->Label(-image=> $window->Photo(-file=>$image),
                          -background=>"white");
