--- /usr/lib/systeminstaller/SystemInstaller/Package.pm.orig    2003-04-11 14:09:03.000000000 -0700
+++ /usr/lib/systeminstaller/SystemInstaller/Package.pm 2004-10-20 07:48:46.000000000 -0700
@@ -72,7 +72,8 @@
 
         
         # Display the line count for GUI status bars.
-        my $linecount = ((scalar(@pkglist) * 2) + $outlines);
+        # 20% is the "slop" from the rpm installs, e.g., scriptlet output, &etc.
+        my $linecount = (1.2*scalar(@pkglist) + $outlines);
         &verbose("Expected lines of output: $linecount");
 
         &verbose("Finding package install files.");
